File: frontend/src/report.ts
import type { AssertionDefinition, AssertionFailureCounts, BranchRouteMetrics, ExecutionPlan, Header, LoadProfile, MetricsBatch, QualityGate, RequestStepMetrics, ScenarioStep, StartRequest, StatusCodeCount } from "./types";
import type { MetricHistoryPoint } from "./metricHistory";
import { isSensitiveHeaderName, isSensitiveHeaderValue, redactHeaders, redactSensitiveURL } from "./secretSanitization";
import type { BaselineComparison } from "./reportBaseline";
import { evaluateQualityGate, normalizeQualityGate } from "./reportQuality";
import type { QualityGateResult } from "./reportQuality";

export { applyBaselineComparison, purgeLegacyRunBaselines } from "./reportBaseline";
export type { BaselineComparison } from "./reportBaseline";
export type { QualityGateResult } from "./reportQuality";

export type RunReport = {
  schemaVersion: 9;
  generatedAtUnixMs: number;
  run: {
    targetUrl: string;
    method: string;
    startedAtUnixMs: number;
    finishedAtUnixMs: number;
    elapsedMs: number;
    batchCount: number;
  };
  config: ReportConfig;
  summary: ReportSummary;
  qualityGate: QualityGateResult;
  baseline: BaselineComparison;
  timeline: ReportTimelinePoint[];
};

export type RunReportMetrics = {
  finalBatch: MetricsBatch | null;
  timeline: readonly MetricHistoryPoint[];
  batchCount: number;
};

type ReportConfig = {
  virtualUsers: number;
  durationMs: number;
  requestTimeoutMs: number;
  maxConnsPerHost: number;
  readBufferSize: number;
  writeBufferSize: number;
  maxResponseBytes: number;
  latencySampleRate: number;
  rateLimitRps: number;
  rampUpMs: number;
  profile?: LoadProfile;
  batchIntervalMs: number;
  qualityGate: QualityGate;
  headers: Header[];
  bodyBytes: number;
  scenarioSteps: ReportScenarioStep[];
  executionPlan?: ExecutionPlan;
};

type ReportScenarioStep = {
  id?: string;
  name?: string;
  kind: ScenarioStep["kind"];
  url?: string;
  method?: string;
  headers?: Header[];
  bodyBytes?: number;
  captures?: Array<{ name: string; path: string; scope: string; onStatus: string }>;
  delayMs?: number;
  expectedStatus?: string;
  assertion?: AssertionDefinition;
};

type ReportSummary = {
  averageRps: number;
  totalRequests: number;
  successRequests: number;
  failedRequests: number;
  successRate: number;
  failureRate: number;
  droppedIterations: number;
  latencySamples: number;
  effectiveLatencySampleRate: number;
  latencyPercentileErrorBoundPct: number;
  latencyMs: {
    avg: number;
    min: number;
    max: number;
    p95: number;
    p99: number;
    p999: number;
  };
  failures: {
    timeout: number;
    dns: number;
    tls: number;
    connRefused: number;
    other: number;
    assertions: number;
    assertionTypes: AssertionFailureCounts;
    captures: number;
    templates: number;
  };
  bytes: {
    read: number;
    written: number;
  };
  statusCodes: StatusCodeCount[];
  requestSteps: RequestStepMetrics[];
  branchRoutes: BranchRouteMetrics[];
};

type ReportTimelinePoint = {
  timestampUnixMs: number;
  rps: number;
  total: number;
  failed: number;
  p95LatencyMs: number;
  p99LatencyMs: number;
};

export function buildRunReport(request: StartRequest, metrics: RunReportMetrics): RunReport {
  const finalBatch = metrics.finalBatch ?? emptyBatch();
  const runtimeSecretValues = request.runtimeSecretValues ?? [];
  const runLatency = finalBatch.runLatency;
  const qualityGate = normalizeQualityGate(request.qualityGate);
  const startedAtUnixMs = validTimestamp(finalBatch.startedAtUnixMs)
    ? finalBatch.startedAtUnixMs
    : metrics.timeline[0]?.timestampUnixMs ?? finalBatch.timestampUnixMs;
  const finishedAtUnixMs = finalBatch.timestampUnixMs;
  const elapsedMs = Math.max(0, finishedAtUnixMs - startedAtUnixMs);
  const total = finalBatch.total;
  const averageRps = elapsedMs > 0 ? total / (elapsedMs / 1_000) : 0;

  return {
    schemaVersion: 9,
    generatedAtUnixMs: Date.now(),
    run: {
      targetUrl: redactReportText(request.config.url, runtimeSecretValues),
      method: request.config.method,
      startedAtUnixMs,
      finishedAtUnixMs,
      elapsedMs,
      batchCount: metrics.batchCount,
    },
    config: {
      virtualUsers: request.config.virtualUsers,
      durationMs: request.config.durationMs,
      requestTimeoutMs: request.config.requestTimeoutMs,
      maxConnsPerHost: request.config.maxConnsPerHost,
      readBufferSize: request.config.readBufferSize,
      writeBufferSize: request.config.writeBufferSize,
      maxResponseBytes: request.config.maxResponseBytes,
      latencySampleRate: request.config.latencySampleRate,
      rateLimitRps: request.config.rateLimitRps,
      rampUpMs: request.config.rampUpMs,
      profile: request.config.profile ? {
        ...request.config.profile,
        stages: request.config.profile.stages.map((stage) => ({ ...stage })),
      } : undefined,
      batchIntervalMs: request.batchIntervalMs,
      qualityGate,
      headers: redactReportHeaders(request.config.headers, runtimeSecretValues),
      bodyBytes: byteLength(request.config.body),
      scenarioSteps: request.config.scenarioSteps.map((step) => redactScenarioStep(step, runtimeSecretValues)),
      executionPlan: request.config.executionPlan ? cloneExecutionPlan(request.config.executionPlan) : undefined,
    },
    summary: {
      averageRps,
      totalRequests: total,
      successRequests: finalBatch.success,
      failedRequests: finalBatch.failed,
      successRate: ratio(finalBatch.success, total),
      failureRate: ratio(finalBatch.failed, total),
      droppedIterations: finalBatch.droppedIterations ?? 0,
      latencySamples: runLatency.samples,
      effectiveLatencySampleRate: ratio(runLatency.samples, total),
      latencyPercentileErrorBoundPct: finalBatch.latencyPercentileErrorBoundPct,
      latencyMs: {
        avg: runLatency.avgMs,
        min: runLatency.minMs,
        max: runLatency.maxMs,
        p95: runLatency.p95Ms,
        p99: runLatency.p99Ms,
        p999: runLatency.p999Ms,
      },
      failures: {
        timeout: finalBatch.timeout,
        dns: finalBatch.dns,
        tls: finalBatch.tls,
        connRefused: finalBatch.connRefused,
        other: finalBatch.otherErrors,
        assertions: finalBatch.assertionsFailed,
        assertionTypes: normalizedAssertionFailures(finalBatch.assertionFailuresByType),
        captures: finalBatch.captureFailures,
        templates: finalBatch.templateFailures,
      },
      bytes: {
        read: finalBatch.bytesRead,
        written: finalBatch.bytesWritten,
      },
      statusCodes: finalBatch.statusCodes.map((status) => ({
        code: status.code,
        count: status.count,
      })),
      requestSteps: (finalBatch.stepMetrics ?? []).map((step) => redactRequestStepMetrics(step, runtimeSecretValues)),
      branchRoutes: (finalBatch.branchMetrics ?? []).map((route) => ({ ...route })),
    },
    qualityGate: evaluateQualityGate(qualityGate, finalBatch, averageRps),
    baseline: {
      key: baselineKey(request),
      verdict: "new-baseline",
      previousRunUnixMs: null,
      deltas: null,
    },
    timeline: metrics.timeline.map((point) => ({ ...point })),
  };
}

function baselineKey(request: StartRequest) {
  const runtimeSecretValues = request.runtimeSecretValues ?? [];
  const scenarioSignature = request.config.scenarioSteps
    .map((step) => [
      step.kind,
      step.method ?? "",
      canonicalURL(redactRuntimeSecrets(step.url ?? "", runtimeSecretValues)),
      step.delayMs ?? "",
      step.expectedStatus ?? "",
      JSON.stringify(step.assertion ? redactAssertion(step.assertion, runtimeSecretValues) : null),
    ].join(":"))
    .join("|");
  const executionSignature = JSON.stringify(request.config.executionPlan ?? null);
  return stableHash([
    request.config.method,
    canonicalURL(redactRuntimeSecrets(request.config.url, runtimeSecretValues)),
    scenarioSignature,
    executionSignature,
  ].join("|"));
}

function cloneExecutionPlan(plan: ExecutionPlan): ExecutionPlan {
  return {
    schemaVersion: 1,
    entryStepId: plan.entryStepId,
    steps: plan.steps.map((step) => ({
      ...step,
      routes: step.routes?.map((route) => ({ ...route })),
    })),
  };
}

function canonicalURL(rawURL: string) {
  const sanitizedURL = redactSensitiveURL(rawURL);
  try {
    const url = new URL(sanitizedURL);
    url.hash = "";
    return url.toString();
  } catch {
    return sanitizedURL.trim();
  }
}

function stableHash(value: string) {
  let hash = 5381;
  for (let index = 0; index < value.length; index += 1) {
    hash = ((hash << 5) + hash) ^ value.charCodeAt(index);
  }
  return `b${(hash >>> 0).toString(36)}`;
}

export function downloadRunReport(report: RunReport) {
  const blob = new Blob([`${JSON.stringify(report, null, 2)}\n`], { type: "application/json" });
  const url = URL.createObjectURL(blob);
  const anchor = document.createElement("a");
  anchor.href = url;
  anchor.download = reportFilename(report);
  anchor.click();
  URL.revokeObjectURL(url);
}

function redactScenarioStep(step: ScenarioStep, runtimeSecretValues: readonly string[]): ReportScenarioStep {
  if (step.kind === "request") {
    return {
      id: step.id,
      name: step.name ? redactReportText(step.name, runtimeSecretValues) : undefined,
      kind: step.kind,
      url: redactReportText(step.url ?? "", runtimeSecretValues),
      method: step.method,
      headers: redactReportHeaders(step.headers ?? [], runtimeSecretValues),
      bodyBytes: byteLength(step.body ?? ""),
      captures: (step.captures ?? []).map((capture) => ({
        name: capture.name,
        path: capture.path,
        scope: capture.scope ?? "iteration",
        onStatus: capture.onStatus ?? "success",
      })),
    };
  }
  if (step.kind === "delay") {
    return {
      id: step.id,
      name: step.name ? redactRuntimeSecrets(step.name, runtimeSecretValues) : undefined,
      kind: step.kind,
      delayMs: step.delayMs,
    };
  }
  return {
    id: step.id,
    name: step.name ? redactRuntimeSecrets(step.name, runtimeSecretValues) : undefined,
    kind: step.kind,
    expectedStatus: step.expectedStatus,
    assertion: step.assertion ? redactAssertion(step.assertion, runtimeSecretValues) : undefined,
  };
}

function redactAssertion(
  assertion: AssertionDefinition,
  runtimeSecretValues: readonly string[],
): AssertionDefinition {
  const sensitiveHeader = assertion.type === "header" && (
    isSensitiveHeaderName(assertion.headerName ?? "") || isSensitiveHeaderValue(assertion.expected ?? "")
  );
  const sensitiveJSON = assertion.type === "json" &&
    (assertion.valueType ?? "string") === "string" &&
    (isSensitiveHeaderName(assertion.jsonPath ?? "") || isSensitiveHeaderValue(assertion.expected ?? ""));
  if (sensitiveHeader || sensitiveJSON) {
    return { ...assertion, expected: assertion.expected ? "<redacted>" : assertion.expected };
  }
  return {
    ...assertion,
    expected: assertion.expected === undefined
      ? undefined
      : redactRuntimeSecrets(assertion.expected, runtimeSecretValues),
  };
}

function normalizedAssertionFailures(counts: AssertionFailureCounts | undefined): AssertionFailureCounts {
  return {
    status: counts?.status ?? 0,
    header: counts?.header ?? 0,
    json: counts?.json ?? 0,
    responseLatency: counts?.responseLatency ?? 0,
    stepLatency: counts?.stepLatency ?? 0,
    countOnly: counts?.countOnly ?? 0,
  };
}

function redactRequestStepMetrics(
  step: RequestStepMetrics,
  runtimeSecretValues: readonly string[],
): RequestStepMetrics {
  return {
    ...step,
    name: redactReportText(step.name, runtimeSecretValues),
    runLatency: { ...step.runLatency },
    statusCodes: step.statusCodes.map((status) => ({ ...status })),
    ...(step.assertionFailuresByType
      ? { assertionFailuresByType: normalizedAssertionFailures(step.assertionFailuresByType) }
      : {}),
  };
}

function redactReportHeaders(headers: Header[], runtimeSecretValues: readonly string[]) {
  return redactHeaders(headers).map((header) => ({
    name: redactRuntimeSecrets(header.name, runtimeSecretValues),
    value: redactRuntimeSecrets(header.value, runtimeSecretValues),
  }));
}

function redactReportText(value: string, runtimeSecretValues: readonly string[]) {
  return redactRuntimeSecrets(redactSensitiveURL(value), runtimeSecretValues);
}

function redactRuntimeSecrets(value: string, runtimeSecretValues: readonly string[]) {
  let redacted = value;
  for (const secret of runtimeSecretValues) {
    if (secret) {
      redacted = redacted.split(secret).join("<redacted>");
    }
  }
  return redacted;
}

function byteLength(value: string) {
  return new TextEncoder().encode(value).length;
}

function ratio(value: number, total: number) {
  return total > 0 ? value / total : 0;
}

function validTimestamp(value: number) {
  return Number.isFinite(value) && value > 0;
}

function emptyBatch(): MetricsBatch {
  const now = Date.now();
  return {
    timestampUnixMs: now,
    startedAtUnixMs: now,
    running: false,
    rps: 0,
    total: 0,
    success: 0,
    failed: 0,
    timeout: 0,
    dns: 0,
    tls: 0,
    connRefused: 0,
    otherErrors: 0,
    assertionsFailed: 0,
    captureFailures: 0,
    templateFailures: 0,
    droppedIterations: 0,
    latencyPercentileErrorBoundPct: 2,
    intervalLatency: {
      samples: 0,
      avgMs: 0,
      p95Ms: 0,
      p99Ms: 0,
      p999Ms: 0,
    },
    runLatency: {
      samples: 0,
      avgMs: 0,
      minMs: 0,
      maxMs: 0,
      p95Ms: 0,
      p99Ms: 0,
      p999Ms: 0,
    },
    bytesRead: 0,
    bytesWritten: 0,
    statusCodes: [],
    stepMetrics: [],
  };
}

function reportFilename(report: RunReport) {
  const host = safeHost(report.run.targetUrl);
  const timestamp = new Date(report.generatedAtUnixMs).toISOString().replace(/[:.]/g, "-");
  return `flowroutine-${host}-${timestamp}.json`;
}

function safeHost(rawURL: string) {
  try {
    return new URL(rawURL).host.replace(/[^a-z0-9.-]+/gi, "-").slice(0, 64) || "report";
  } catch {
    return "report";
  }
}

File: internal/headless/report.go
package headless

import (
	"encoding/json"
	"encoding/xml"
	"fmt"
	"strconv"
	"time"

	"flowroutine/internal/bridge"
)

const ReportSchemaVersion = 1

type RunReport struct {
	SchemaVersion     int                      `json:"schemaVersion"`
	GeneratedAtUnixMS int64                    `json:"generatedAtUnixMs"`
	Scenario          ReportScenario           `json:"scenario"`
	Run               ReportRun                `json:"run"`
	Preflight         bridge.PreflightResponse `json:"preflight"`
	Summary           ReportSummary            `json:"summary"`
	QualityGate       QualityGateResult        `json:"qualityGate"`
}

type ReportScenario struct {
	ID       string `json:"id"`
	Name     string `json:"name"`
	Revision int    `json:"revision"`
}

type ReportRun struct {
	StartedAtUnixMS  int64 `json:"startedAtUnixMs"`
	FinishedAtUnixMS int64 `json:"finishedAtUnixMs"`
	ElapsedMS        int64 `json:"elapsedMs"`
	BatchCount       int   `json:"batchCount"`
}

type ReportSummary struct {
	AverageRPS                     float64                         `json:"averageRps"`
	TotalRequests                  uint64                          `json:"totalRequests"`
	SuccessRequests                uint64                          `json:"successRequests"`
	FailedRequests                 uint64                          `json:"failedRequests"`
	SuccessRate                    float64                         `json:"successRate"`
	FailureRate                    float64                         `json:"failureRate"`
	DroppedIterations              uint64                          `json:"droppedIterations"`
	LatencySamples                 uint64                          `json:"latencySamples"`
	EffectiveLatencySampleRate     float64                         `json:"effectiveLatencySampleRate"`
	LatencyPercentileErrorBoundPct float64                         `json:"latencyPercentileErrorBoundPct"`
	LatencyMS                      bridge.CumulativeLatencyMetrics `json:"latencyMs"`
	Failures                       ReportFailures                  `json:"failures"`
	Bytes                          ReportBytes                     `json:"bytes"`
	StatusCodes                    []bridge.StatusCodeCount        `json:"statusCodes"`
	RequestSteps                   []bridge.RequestStepMetrics     `json:"requestSteps"`
}

type ReportFailures struct {
	Timeout     uint64                        `json:"timeout"`
	DNS         uint64                        `json:"dns"`
	TLS         uint64                        `json:"tls"`
	ConnRefused uint64                        `json:"connRefused"`
	Other       uint64                        `json:"other"`
	Assertions  uint64                        `json:"assertions"`
	Types       bridge.AssertionFailureCounts `json:"assertionTypes"`
	Captures    uint64                        `json:"captures"`
	Templates   uint64                        `json:"templates"`
}

type ReportBytes struct {
	Read    uint64 `json:"read"`
	Written uint64 `json:"written"`
}

func BuildRunReport(file *ScenarioFile, preflight bridge.PreflightResponse, final bridge.MetricsBatch, batchCount int, generatedAt time.Time) *RunReport {
	startedAt := final.StartedAtUnixMs
	finishedAt := final.TimestampUnixMs
	if startedAt <= 0 || finishedAt < startedAt {
		startedAt = finishedAt
	}
	elapsedMS := finishedAt - startedAt
	averageRPS := 0.0
	if elapsedMS > 0 {
		averageRPS = float64(final.Total) / (float64(elapsedMS) / 1_000)
	}
	gate := NormalizeQualityGate(file.Scenario.QualityGate)
	return &RunReport{
		SchemaVersion:     ReportSchemaVersion,
		GeneratedAtUnixMS: generatedAt.UnixMilli(),
		Scenario: ReportScenario{
			ID:       file.Scenario.ID,
			Name:     file.Scenario.Name,
			Revision: file.Scenario.Revision,
		},
		Run: ReportRun{
			StartedAtUnixMS:  startedAt,
			FinishedAtUnixMS: finishedAt,
			ElapsedMS:        elapsedMS,
			BatchCount:       batchCount,
		},
		Preflight: preflight,
		Summary: ReportSummary{
			AverageRPS:                     averageRPS,
			TotalRequests:                  final.Total,
			SuccessRequests:                final.Success,
			FailedRequests:                 final.Failed,
			SuccessRate:                    ratio(final.Success, final.Total),
			FailureRate:                    ratio(final.Failed, final.Total),
			DroppedIterations:              final.DroppedIterations,
			LatencySamples:                 final.RunLatency.Samples,
			EffectiveLatencySampleRate:     ratio(final.RunLatency.Samples, final.Total),
			LatencyPercentileErrorBoundPct: final.LatencyPercentileErrorBoundPct,
			LatencyMS:                      final.RunLatency,
			Failures: ReportFailures{
				Timeout:     final.Timeout,
				DNS:         final.DNS,
				TLS:         final.TLS,
				ConnRefused: final.ConnRefused,
				Other:       final.OtherErrors,
				Assertions:  final.AssertionsFailed,
				Types:       final.AssertionFailuresByType,
				Captures:    final.CaptureFailures,
				Templates:   final.TemplateFailures,
			},
			Bytes: ReportBytes{
				Read:    final.BytesRead,
				Written: final.BytesWritten,
			},
			StatusCodes:  cloneStatusCodes(final.StatusCodes),
			RequestSteps: cloneRequestSteps(final.StepMetrics),
		},
		QualityGate: EvaluateQualityGate(gate, final, averageRPS),
	}
}

func MarshalJSONReport(report *RunReport) ([]byte, error) {
	data, err := json.MarshalIndent(report, "", "  ")
	if err != nil {
		return nil, fmt.Errorf("encode JSON report: %w", err)
	}
	return append(data, '\n'), nil
}

func MarshalJUnitReport(report *RunReport) ([]byte, error) {
	suite := junitTestSuite{
		Name:      "FlowRoutine " + report.Scenario.Name,
		Timestamp: time.UnixMilli(report.GeneratedAtUnixMS).UTC().Format(time.RFC3339),
		Time:      secondsString(report.Run.ElapsedMS),
	}
	suite.Cases = append(suite.Cases, junitTestCase{
		Name:      "scenario_execution",
		ClassName: "flowroutine." + report.Scenario.ID,
		Time:      secondsString(report.Run.ElapsedMS),
	})
	for _, check := range report.QualityGate.Checks {
		caseResult := junitTestCase{
			Name:      check.Name,
			ClassName: "flowroutine." + report.Scenario.ID + ".quality_gate",
		}
		message := fmt.Sprintf("actual %g %s threshold %g", check.Actual, check.Operator, check.Threshold)
		switch check.Status {
		case QualityGateFail:
			caseResult.Failure = &junitFailure{Message: message, Body: message}
			suite.Failures++
		case QualityGateInsufficient:
			caseResult.Skipped = &junitSkipped{Message: fmt.Sprintf("%s; samples %d, minimum %d", message, check.Samples, check.MinimumSamples)}
			suite.Skipped++
		}
		suite.Cases = append(suite.Cases, caseResult)
	}
	suite.Tests = len(suite.Cases)
	return marshalJUnit(junitTestSuites{
		Tests:    suite.Tests,
		Failures: suite.Failures,
		Skipped:  suite.Skipped,
		Time:     suite.Time,
		Suites:   []junitTestSuite{suite},
	})
}

func MarshalJUnitError(scenarioName string, scenarioID string, kind string, runError error) ([]byte, error) {
	if scenarioName == "" {
		scenarioName = "scenario"
	}
	if scenarioID == "" {
		scenarioID = "unknown"
	}
	message := kind
	if runError != nil {
		message += ": " + runError.Error()
	}
	suite := junitTestSuite{
		Name:     "FlowRoutine " + scenarioName,
		Tests:    1,
		Failures: 1,
		Cases: []junitTestCase{{
			Name:      kind,
			ClassName: "flowroutine." + scenarioID,
			Failure:   &junitFailure{Message: message, Body: message},
		}},
	}
	return marshalJUnit(junitTestSuites{Tests: 1, Failures: 1, Suites: []junitTestSuite{suite}})
}

func ratio(numerator uint64, denominator uint64) float64 {
	if denominator == 0 {
		return 0
	}
	return float64(numerator) / float64(denominator)
}

func cloneStatusCodes(values []bridge.StatusCodeCount) []bridge.StatusCodeCount {
	if len(values) == 0 {
		return []bridge.StatusCodeCount{}
	}
	return append([]bridge.StatusCodeCount(nil), values...)
}

func cloneRequestSteps(values []bridge.RequestStepMetrics) []bridge.RequestStepMetrics {
	if len(values) == 0 {
		return []bridge.RequestStepMetrics{}
	}
	cloned := append([]bridge.RequestStepMetrics(nil), values...)
	for index := range cloned {
		cloned[index].StatusCodes = cloneStatusCodes(cloned[index].StatusCodes)
	}
	return cloned
}

func secondsString(milliseconds int64) string {
	return strconv.FormatFloat(float64(milliseconds)/1_000, 'f', 3, 64)
}

type junitTestSuites struct {
	XMLName  xml.Name         `xml:"testsuites"`
	Tests    int              `xml:"tests,attr"`
	Failures int              `xml:"failures,attr"`
	Skipped  int              `xml:"skipped,attr"`
	Time     string           `xml:"time,attr,omitempty"`
	Suites   []junitTestSuite `xml:"testsuite"`
}

type junitTestSuite struct {
	Name      string          `xml:"name,attr"`
	Tests     int             `xml:"tests,attr"`
	Failures  int             `xml:"failures,attr"`
	Skipped   int             `xml:"skipped,attr"`
	Time      string          `xml:"time,attr,omitempty"`
	Timestamp string          `xml:"timestamp,attr,omitempty"`
	Cases     []junitTestCase `xml:"testcase"`
}

type junitTestCase struct {
	Name      string        `xml:"name,attr"`
	ClassName string        `xml:"classname,attr"`
	Time      string        `xml:"time,attr,omitempty"`
	Failure   *junitFailure `xml:"failure,omitempty"`
	Skipped   *junitSkipped `xml:"skipped,omitempty"`
}

type junitFailure struct {
	Message string `xml:"message,attr"`
	Body    string `xml:",chardata"`
}

type junitSkipped struct {
	Message string `xml:"message,attr"`
}

func marshalJUnit(suites junitTestSuites) ([]byte, error) {
	data, err := xml.MarshalIndent(suites, "", "  ")
	if err != nil {
		return nil, fmt.Errorf("encode JUnit report: %w", err)
	}
	output := append([]byte(xml.Header), data...)
	return append(output, '\n'), nil
}
